### ISSUE TITLE
Monitoring of value head sharpness

### ISSUE THREAD

**Issue by Ishinoshita** [leela-zero/leela-zero#2165]
LZ training server switched to -m 30 on 10-Jan-2019, IIRC. 'All' networks up to LZ200 were trained only with selfplay games ran with -m 999 ('all' = the last 9 months at least; I didn't retrieved the date of the switch to -m 999).

LZ201 is the first network that has (possibly ?) already been trained with some -m 30 selfplay games in the training window. *Edit: untrue, there has been for a while - m 30 games produced by BigRedII, in variable proportion over time but at least up to 15% of the total self-play games by period.*
With a 500k window, I assume that the switch from -m999 to -m 30 will be quite smoothed. Anyway, I'm eager to see if that changes LZ behaviour in some measurable way.

@gcp raised concerns that -m 30 setting may lead to a too "stratified" value head, as exemplified by Elf's one. I wanted to devise a simple way to compare different networks value head sharpness. Late results I saw in this respect on Discord LZ/Handicap were based on the "color input planes hack", but most resulting curves were not monotonic at all, sometimes with multiple bounces, and anyway far from sigmoids and difficult to compare between themselves. Also, I wanted something really simple to do (no compiling, no coding).

So, I opted for a very basic approach: Komi=7.5, Black to move, Value = NN Eval (Handicap stones), using the following test positions suite, i.e. using black stones, resp. white stones to achieve some sort of algebric handicap H:
![image](https://user-images.githubusercontent.com/37498331/51500428-315a9800-1dce-11e9-82f5-60f160e44bf1.png)
[Handicap sgf suite.zip](https://github.com/gcp/leela-zero/files/2780660/Handicap.sgf.suite.zip)

*NB1: this does not correspond to the usual H2, H3 , etc... in handicap games, where White komi is reduced to 0.5pts.*

*NB2: I aimed at quick & dirty estimation, so I overlooked the question of random seed and 8 symetries/rotations (though some tests I did showed that the NN Eval differences between runs were low enough (range of less than 3% in relative value) with respect to the distribution of NN Eval vs Handicap or vs network generation).*

I got the following curves for various generation of networks:

![image](https://user-images.githubusercontent.com/37498331/51500449-43d4d180-1dce-11e9-8463-9f91260f267b.png)

*  Remarkably, networks perfectly agree on empty board value across generations!
*  Overall, there is a trend towards sharpness with more training. Recent nets (LZ200) are much more opinionated than older ones (LZ150).
*  Surprisingly, this trend is not monotonic with training steps and exhibit several turnaround. There was a trend towards more smoothnes between LZ150 and LZ170, then a sudden reversal/jump with LZ175, up to a maximum sharpness around LZ190. Followed by more smoothness again in recent networks.

I can just elaborate on the reasons behind these changes: more training games, more blocks/filters, Elfv0/Elfv1 games addition in variable proportion in the training window, proportion of no-resign games. *Edit: and proportion of - m 30 self-play games from BRII, promotions of @bjiyxo nets (trained with slightly different method),  etc....*. There might also be some artifacts due to my sampling of 1 network out of 5. Or this method sucks for other reasons...

I will update this graph in a while with next generations of networks, trained with more and more, then solely with -m 30 games. This is very cheap study.

Hopefully, we will not end-up there!:
![image](https://user-images.githubusercontent.com/37498331/51500465-53541a80-1dce-11e9-8d6b-476bec8390bb.png)

@amj BTW, minigo is also using "-m 30" since the begining IIRC. , what does MG's value head curve looks like for v15, with this method ?

**Comment by amj** [comment]
to directly answer your question, we haven't measured :)  But the convert_minigo_weights.py script works, so you can run the same analysis and put MG on the same plots, if you wished :)

To less directly answer your question: I am not sure how much whole-game temperature is the cause of value sharpness; and i'm not sure we use the same terms for 'value sharpness'.  It seems very reasonable for ELF to say 'if black plays a star point, and white passes, then it's almost impossible for white to win.'  This seems to be exactly what the training data should cause to happen, and i'm not sure it's even undesirable:  I expect if ELF played itself 1000 times where white passes for its first move, all randomness turned off, then i would not be surprised if ELF won < 10 of them.  And in some sense, that's exactly what that curve is saying.

Not only that, but i think that's the desired result: if our NN is getting good, we would expect it to 'convert' favorable positions -- and getting a free move definitely counts as a favorable position :)  So getting this 'correct' result, i.e., the sharpness shown here by ELF, is not necessarily *wrong*, although it might be less useful for our human purposes.

A different kind of sharpness is, say, black opens at the 3-4 point and ELF says 'oh gosh now w has 85% chance' or whatever.  That kind of sharpness seems probably indicative of the overall RL process being less correct and not having enough diversity in training data.

Lastly, i think whole game temperature reduces value confidence by essentially toggling the outcome of the game (from some hypothetical perfect play).  @sethtroisi measured how much 'toggled data' a net can withstand before its performance degrades -- but degraded performance is not the same as the sort of 'opinionated sharpness'.

anyway hope this helps :)

**Comment by barrtgt** [comment]
Having a stronger network will give a more sensitive value estimation. There is no getting around that. It appears -m 999 did slow it down from being so sharp, but there was no clear strength improvement, at least according to the cloudygo chart. Is there a better option to improving networks? Personally, I think if we have a flatter distribution like we have now, it would be beneficial to have more visits per move. It also depends on what people want Leela to provide. Would people rather want a stronger ai that only provides a few good moves and is a little less useful for winrate estimation later in games, but gives a stronger endgame, or one that provides more move options, is weaker, a little better winrate estimation later in games, but worse endgame?

**Comment by Ishinoshita** [comment]
I agree that to "strongly" direct the search and yield strength, the value head must be sharp. I understand the tradeoff. By saying "Hopefully, we will not end-up there", I refered to @gcp [mentioning](https://github.com/gcp/leela-zero-server/pull/216#issuecomment-454399618) that he does not want to just to make a stronger Elf. So he may just remove the plug on -m 30. Was in favor of -m 100.

So, my point was not to raise an issue or debate what sharpness is best, just to say that this simple method  allows to visualize in some way the value head sharpness and to compare between networks. Although  people often alledge 'higher sharpness', 'change in sharpness', and although this method is a very straightforwards/obvious way to characterize the value head, I don't recall seing such plots.

@amj Thanks! I'll answer my own question myself as soon as I can locate a converted weight for MG v15.

**Comment by roy7** [comment]
@Ishinoshita The convertor is in https://github.com/gcp/leela-zero/tree/next/training/minigo

**Comment by Splee99** [comment]
Does this mean that early networks are better at handicap games?

**Comment by Ishinoshita** [comment]
I haven't tried, so it's purely speculative, but it might be that, particularly at high handicap, a weaker network (in even games), with a smoother value head, could do better than a stronger one with a sharp value head.

**Comment by Ishinoshita** [comment]
@sethtroisi @amj Thanks for the converted weights! I've plotted the evolution of minigo v15 value head (average over 8 symetries) for different handicap stones, during the training. Just out of curiousity.

Global trend towards sharpness but not monotonic at all during the training. For each particular handicap setting, value head often exibit one or more turnarounds. Just tells us that the network journey during training is something complex...

*NB1: Excel smoothed curves are of course overstatement. No idea in practice if the turnarounds are smooth or sharp during training.*
*NB2: I double-checked network#800 behaviour. Black turn NN Eval is really higher when there are already two White stones (12%) than when there is only one (10%) ! ;-)* 

![image](https://user-images.githubusercontent.com/37498331/51805309-2b503580-226c-11e9-9f01-a71d30c62920.png)

![image](https://user-images.githubusercontent.com/37498331/51805318-3c00ab80-226c-11e9-9d92-4e6d77cea30f.png)

Comparison of LZ200, Minigo v15_1000, _0900 et ELFv1:
![image](https://user-images.githubusercontent.com/37498331/51805526-aa466d80-226e-11e9-8b7b-dca0e15762d8.png)

*NB: as discussed by @amj on Discord, these values does not tell how the search will estimate those positions. For instance, for net v15_1000, even if empty board NN Eval is unfavorable to Black (42%), win rate quicky converges ~50%-51% after a few hundreds playouts.*

**Comment by sethtroisi** [comment]
@Ishinoshita Thanks for doing these calculations.

**Comment by Ishinoshita** [comment]
Well, it's hard for me to draw any meaningful conclusion...beyond the fact that MG value doesn't look prima facie much sharper or smoother than Elf1, and that sharpness may significantly vary from net to net, as was observed with LZ with variable komi trick. 

I also note that the value head for H0 is quite stable, kind of fixed point of value head training, whereas values for other handicaps continue to 'freely' oscillate even late in the training, kind of degrees of freedom, as accurate evaluation of largely out of balance positions doesn't care much, as long as the curve keeps monotonic. 

But even this is probably overstatement, as these values are just the result of net generalizing over positions never encountered in the training data. Oppositely, slightly out of balance positions are probably much more represented in training data and their accurate evaluation is of more importance.

**Comment by nerai** [comment]
I would like to kindly ask about an update in this topic. There have been some developments and new nets that may be interesting to compare.
